FIX Reduces memory usage of `PCA.transform`

Micky774 opened the pull request:
#### Reference Issues/PRs
Fixes #11102

#### What does this implement/fix? Explain your changes.
This PR implements an iterative algorithm for calculating variance for sufficiently large data (`X.shape[0] > 1000`), thereby reducing the peak memory footprint, regardless of choice of `copy` argument.

#### Any other comments?
- **Currently failing two tests** due to relative error of approximately `1.4e-5`
- There is a negligible (0s-4s) slowdown using the iterative method on small/medium datasets (`num_features < 6400`) due to the induced overhead, however it is incredibly close in speed to the normal method on large datasets.

Micky774 committed:
Initial in-place variance calculation implementation

Micky774 committed:
Implemented iterative variance calculation

Micky774 committed:
Merge branch 'main' into pca_mem

Micky774 committed:
Merge branch 'main' into pca_mem

jeremiedbb commented:
I'm not convinced about that kind of optimisation. It adds a lot of complexity in terms of maintenance and readability to get around a flow in a single numpy function. I think it would be much better that the memory issue is fixed directly in numpy. There are discussions and even a PR about that https://github.com/numpy/numpy/issues/13199, https://github.com/numpy/numpy/pull/13263. The PR seems a bit stalled but maybe we can see if they are still willing to finish it.

Micky774 committed:
Added new test, introduced in-place var calculation

- Created a new method in `extmath.py` for calculating feature-wise
in-place variance
- PCA now uses in-place feature variance calculation to avoid mem spike
- Added new test to check for correctness of internal `total_var`
calculation
- Fixed existing whitening test by changing `_X`->`_X.copy` to account
for the `copy=False` semantics of accepting in-place changes in `X`

Micky774 commented:
@jeremiedbb Looks like it's not happening, at least on the numpy side of things :(

I agree regarding the induced complexity of the iterative calculation -- actually it looks like it was also very inaccurate, so I'm no longer considering that option.

Instead, I have added my `in_place_feature_var` method to `extmath.py` and included a new test for correctness. Currently when the user flags `copy=False` they agree to allow the input `X` to be mutated, therefore this in-place calculation is acceptable in both cases. I agree it's less than ideal, but I think it's the best option we have right now. It currently passes the existing tests\* and matches the `np.var(X, ddof=1, axis=0)` function.

The in-place method results in essentially no new allocated memory, and runs faster than the old method.

\* I did have to modify the `test_whitening` in `test_pca.py` but I believe it is a proper correction. Essentially, they omitted a copy call to one of the inputs to `fit`.

cc: @thomasjpfan

Micky774 committed:
Added changelog entry

Micky774 committed:
Removed empty line

Micky774 committed:
Added test parameterization for `copy`

Micky774 committed:
Merge branch 'main' into pca_mem

jeremiedbb commented on sklearn/decomposition/_pca.py:
```diff
@@ -619,12 +619,12 @@ def _fit_truncated(self, X, n_components, svd_solver):
         self.explained_variance_ = (S**2) / (n_samples - 1)
-        total_var = np.var(X, ddof=1, axis=0)
-        self.explained_variance_ratio_ = self.explained_variance_ / total_var.sum()
+        total_var = in_place_feature_var(X).sum()
```
Turns out X is already centered at this point, see l.593. So we can avoid this extra computation. Based on that I don't think it's necessary to define a function for these 3 lines.

```suggestion
        np.square(X, out=X)
        np.sum(X, axis=0, out=X[0])
        feature_var = X[0] / (X.shape[0] - 1)
        total_var = feature_var.sum()
```

jeremiedbb reviewed (commented):
@Micky774 I find this solution much cleaner !

I still think it's the role of numpy to implement a low memory var, so maybe add a comment to just say that it's a workaround because numpy doesn't have that yet.

Micky774 committed:
Merge branch 'main' into pca_mem

Micky774 committed:
Moved calc out of `extmath.py` and directly into `_pca.py`

Micky774 commented:
I've moved the remainder of the in-place variance calculation into `_pca.py` and removed the now unnecessary function from `extmath.py`. I also included the workaround comment in the code so hopefully there'll be a more appropriate replacement later :)

jeremiedbb reviewed (approved):
LGTM. Thanks @Micky774 !

thomasjpfan reviewed (approved):
LGTM
